doc/man/pop3d.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH POP3D 1 "November 10th, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
pop3d \- RFC1939 compliant POP3 Daemon. 
.SH SYNOPSIS
.B pop3d
.RI [ options ]
.SH DESCRIPTION
This manual page documents briefly the
.B pop3d
command.
This man page is not intended to be complete.
Instead, it has documentation in the GNU Info format; see below.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.TP
.B \-d, \-\-daemon=MAXCHILDREN
Runs in daemon mode with a maximum of MAXCHILDREN child processes.
.TP
.B \-i, \-\-inetd
Runs in inetd mode (default).
.TP
.B \-p, \-\-port=PORT
Specifies port to listen on, implies -d defaults to 110, which need not be specified.
.TP
.B \-t, \-\-timeout=TIMEOUT
Sets idle timeout to TIMEOUT seconds TIMEOUT default is 600 (10 minutes).
.SH SEE ALSO
.BR RFC1939.
.br
The programs are documented fully by
.IR "mailutils" ,
available via the Info system.
.SH AUTHOR
This manual page was written by Jeff Bailey <jbailey@gnu.org>.
Copyright (C) 2001 Free Software Foundation.  This document is
in the PUBLIC DOMAIN.


NEWS
GNU mailutils NEWS -- history of user-visible changes.  2003-02-12
Copyright (C) 2002, 2003 Free Software Foundation, Inc.
See the end for copying conditions.

Please send mailutils bug reports to <bug-mailutils@gnu.org>.

Version 0.3:

* General:

** Added NLS support. Currently Catalan and Spanish translations
are available.

** Fixed compilation with the newer versions of Berkeley DB.

* libmailbox:

** Functions for generating References: and In-Reply-To: headers as
per RFC 2822.

** A set of functions for sequential access to streams.

** The function mu_errstring is renamed to mu_strerror for consistency
with the usual practice.

** Added initial implementation of message header encoding and decoding
functions (RFC 2047).

* libsieve:

** The syntax of multiline strings is extended. It is possible
to use arbitrary ending delimiter instead of the default dot,
and to strip off the leading tabs, which allows for natural
indentation of the multiline strings.

** Added support for relational tests according to RFC 3431.

** Fixed incorrect handling of multiple addresses in `address'
tests.

* imap4d:

** The untagged output was fixed to conform to the RFC 2060. Specifically,
the output of STORE command is corrected in accordance with the
rfc2060-errata document; the EXIST and RECENT untagged are always
returned by EXAMINE, SELECT, and NOOP.

** Added support for TLS and GSASL.

** Implemented LOGINDISABLED capability.

** Implemented IDLE command (RFC 2177).

* pop3d: Added TLS support.

* mail: Fixed error diagnostics and removed startup banner.

* mh: Initial implementation. 


Version 0.2:

* libmailbox: Several bugfixes. The configuration suite allows
  to disable support for any protocol or mailbox format to reduce
  the size of the library.

* Added new utility mailutils-config. This utility prints gcc command
  line options needed for compiling and linknig an application against
  mailutils.

* libsieve: New library. Supports Sieve language as described in
  RFC 3028 and provides a mechanism for dynamic loading of
  user-defined actions, tests and comparators.

* sieve: Completely rewritten using libsieve.

* sieve.scm: Added vacation extension action.

* mail.local: Allows to apply user-defined sieve filters before delivering
  messages.

* mail utility: Added `editheaders' boolean variable. When set, the
  user is allowed to edit headers of the message.
  Message set specification allows negation of search primitives, e.g.
  `from ! subject:/daily/' outputs all messages whose subject
  does not contain the word 'daily'.
  Reply command correctly adds `In-Reply-To' and `References' headers
  to outgoing messages.
  When used with a numeric argument '=' command sets the current
  message to this number.
  New command `setenv' allows to set and examine environment
  variables.
  The semantics of `crt' variable is made compatible with that of
  other implementations.
  Fixed several bugs.

* libmu_scm: Added support for Guile 1.7.0


Version 0.1:

The first release.


----------------------------------------------------------------------
Copyright information:

Copyright (C) 2002 Free Software Foundation, Inc.

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.

Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


pop3d/pop3d.c
/* GNU mailutils - a suite of utilities for electronic mail
   Copyright (C) 1999, 2000, 2001 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

#include "pop3d.h"

/* Save some line space.  */
typedef struct sockaddr_in SA;

/* Declared in <pop3d.h>.  */
mailbox_t mbox;

unsigned int port;
unsigned int timeout;
int state;
char *username;
int ifile;
FILE *ofile;
time_t curr_time;
char *md5shared;
volatile unsigned int children;

/* Number of child processes.  */
unsigned int children = 0;

static struct option long_options[] =
{
  {"daemon", optional_argument, 0, 'd'},
  {"help", no_argument, 0, 'h'},
  {"inetd", no_argument, 0, 'i'},
  {"port", required_argument, 0, 'p'},
  {"timeout", required_argument, 0, 't'},
  {"version", no_argument, 0, 'v'},
  {0, 0, 0, 0}
};

const char *short_options ="d::hip:t:v";

int
main (int argc, char **argv)
{
  struct group *gr;
  static int mode = INTERACTIVE;
  size_t maxchildren = 20;
  int c = 0;
  int status = OK;

  port = 110;			/* Default POP3 port.  */
  timeout = 0;			/* Default timeout of 0.  */

  while ((c = getopt_long (argc, argv, short_options, long_options, NULL))
	 != -1)
    {
      switch (c)
	{
	case 'd':
	  mode = DAEMON;
	  maxchildren = optarg ? strtoul (optarg, NULL, 10) : 10;
	  if (maxchildren <= 0)
	    maxchildren = 10;
	  break;

	case 'h':
	  pop3d_usage (argv[0]);
	  break;

	case 'i':
	  mode = INTERACTIVE;
	  break;

	case 'p':
	  mode = DAEMON;
	  port = strtoul (optarg, NULL, 10);
	  break;

	case 't':
	  timeout = strtoul (optarg, NULL, 10);
	  break;

	case 'v':
	  printf (IMPL " ("PACKAGE " " VERSION ")\n");
	  exit (0);
	  break;

	default:
	  break;
	}
    }

  /* First we want our group to be mail so we can access the spool.  */
  gr = getgrnam ("mail");
  if (gr == NULL)
    {
      perror ("Error getting group");
      exit (1);
    }

  if (setgid (gr->gr_gid) == -1)
    {
      perror ("Error setting group");
      exit (1);
    }

  /* Register the desire formats. We only need Mbox mail format.  */
  {
    list_t bookie;
    registrar_get_list (&bookie);
    /* list_append (bookie, mbox_record); */
    list_append (bookie, path_record);
  }

  /* Set the signal handlers.  */
  signal (SIGINT, pop3d_signal);
  signal (SIGQUIT, pop3d_signal);
  signal (SIGILL, pop3d_signal);
  signal (SIGBUS, pop3d_signal);
  signal (SIGFPE, pop3d_signal);
  signal (SIGSEGV, pop3d_signal);
  signal (SIGTERM, pop3d_signal);
  signal (SIGSTOP, pop3d_signal);
  signal (SIGPIPE, pop3d_signal);

  if (timeout < 600)		/* RFC 1939 says no less than 10 minutes.  */
    timeout = 0;		/* So we'll turn it off.  */

  if (mode == DAEMON)
    pop3d_daemon_init ();

  /* Change directory.  */
  chdir ("/");

  /* Set up for syslog.  */
  openlog ("gnu-pop3d", LOG_PID, LOG_MAIL);

  umask (S_IROTH | S_IWOTH | S_IXOTH);	/* 007 */

  /* Actually run the daemon.  */
  if (mode == DAEMON)
    pop3d_daemon (maxchildren);
  /* exit (0) -- no way out of daemon except a signal.  */
  else
    status = pop3d_mainloop (fileno (stdin), fileno (stdout));

  /* Close the syslog connection and exit.  */
  closelog ();
  return (OK != status);
}

/* Sets things up for daemon mode.  */
void
pop3d_daemon_init (void)
{
  pid_t pid;
  unsigned int i;
#define MAXFD 64

  pid = fork ();
  if (pid == -1)
    {
      perror ("fork failed:");
      exit (1);
    }
  else if (pid > 0)
    exit (0);			/* Parent exits.  */

  setsid ();			/* Become session leader.  */

  signal (SIGHUP, SIG_IGN);	/* Ignore SIGHUP.  */

  pid = fork ();
  if (pid == -1)
    {
      perror("fork failed:");
      exit (1);
    }
  else if (pid > 0)
    exit (0);			/* Parent exits.  */

  /* Close inherited file descriptors.  */
  for (i = 0; i < MAXFD; ++i)
    close (i);

#ifdef HAVE_SIGACTION
  {
    struct sigaction act;
    act.sa_handler = pop3d_sigchld;
    sigemptyset (&act.sa_mask);
    act.sa_flags = 0;
    sigaction (SIGCHLD, &act, NULL);
  }
#else
  signal (SIGCHLD, pop3d_sigchld);
#endif
}

/* The main part of the daemon. This function reads input from the client and
   executes the proper functions. Also handles the bulk of error reporting.  */

int
pop3d_mainloop (int infile, int outfile)
{
  int status = OK;
  char *buf, *arg, *cmd;
  struct hostent *htbuf;
  char *local_hostname;

  ifile = infile;
  ofile = fdopen (outfile, "w");
  if (ofile == NULL)
    pop3d_abquit (ERR_NO_OFILE);
  state = AUTHORIZATION;
  curr_time = time (NULL);

  /* FIXME:  Retreive hostname with getpeername() and log.  */
  syslog (LOG_INFO, "Incoming connection opened");

  /* log information on the connecting client */
  {
    struct sockaddr_in cs;
    int len = sizeof cs;
    if (getpeername (infile, (struct sockaddr*)&cs, &len) < 0)
      syslog (LOG_ERR, "can't obtain IP address of client: %s",
	      strerror (errno));
    else
      syslog (LOG_INFO, "connect from %s", inet_ntoa(cs.sin_addr));
  }

  /* Prepare the shared secret for APOP.  */
  local_hostname = malloc (MAXHOSTNAMELEN + 1);
  if (local_hostname == NULL)
    pop3d_abquit (ERR_NO_MEM);

  gethostname (local_hostname, MAXHOSTNAMELEN);
  htbuf = gethostbyname (local_hostname);
  if (htbuf)
    {
      free (local_hostname);
      local_hostname = strdup (htbuf->h_name);
    }

  md5shared = malloc (strlen (local_hostname) + 51);
  if (md5shared == NULL)
    pop3d_abquit (ERR_NO_MEM);

  snprintf (md5shared, strlen (local_hostname) + 50, "<%u.%u@%s>", getpid (),
	    (int)time (NULL), local_hostname);
  free (local_hostname);

  fprintf (ofile, "+OK POP3 Ready %s\r\n", md5shared);

  while (state != UPDATE)
    {
      fflush (ofile);
      status = OK;
      buf = pop3d_readline (ifile);
      cmd = pop3d_cmd (buf);
      arg = pop3d_args (buf);

      if (state == TRANSACTION && !mailbox_is_updated (mbox))
	{
	  static size_t mailbox_size;
	  size_t newsize = 0;
	  mailbox_get_size (mbox, &newsize);
	  /* Did we shrink?  */
	  if (!mailbox_size)
	    mailbox_size = newsize;
	  else (newsize < mailbox_size)
	    pop3d_abquit (ERR_MBOX_SYNC);
	}

      if (strlen (arg) > POP_MAXCMDLEN || strlen (cmd) > POP_MAXCMDLEN)
	status = ERR_TOO_LONG;
      else if (strlen (cmd) > 4)
	status = ERR_BAD_CMD;
      else if (strncasecmp (cmd, "RETR", 4) == 0)
	status = pop3d_retr (arg);
      else if (strncasecmp (cmd, "DELE", 4) == 0)
	status = pop3d_dele (arg);
      else if (strncasecmp (cmd, "USER", 4) == 0)
	status = pop3d_user (arg);
      else if (strncasecmp (cmd, "QUIT", 4) == 0)
	status = pop3d_quit (arg);
      else if (strncasecmp (cmd, "APOP", 4) == 0)
	status = pop3d_apop (arg);
      else if (strncasecmp (cmd, "AUTH", 4) == 0)
	status = pop3d_auth (arg);
      else if (strncasecmp (cmd, "STAT", 4) == 0)
	status = pop3d_stat (arg);
      else if (strncasecmp (cmd, "LIST", 4) == 0)
	status = pop3d_list (arg);
      else if (strncasecmp (cmd, "NOOP", 4) == 0)
	status = pop3d_noop (arg);
      else if (strncasecmp (cmd, "RSET", 4) == 0)
	status = pop3d_rset (arg);
      else if ((strncasecmp (cmd, "TOP", 3) == 0) && (strlen (cmd) == 3))
	status = pop3d_top (arg);
      else if (strncasecmp (cmd, "UIDL", 4) == 0)
	status = pop3d_uidl (arg);
      else if (strncasecmp (cmd, "CAPA", 4) == 0)
	status = pop3d_capa (arg);
      else
	status = ERR_BAD_CMD;

      if (status == OK)
	fflush (ofile);
      else if (status == ERR_WRONG_STATE)
	fprintf (ofile, "-ERR " BAD_STATE "\r\n");
      else if (status == ERR_BAD_ARGS)
	fprintf (ofile, "-ERR " BAD_ARGS "\r\n");
      else if (status == ERR_NO_MESG)
	fprintf (ofile, "-ERR " NO_MESG "\r\n");
      else if (status == ERR_MESG_DELE)
	fprintf (ofile, "-ERR " MESG_DELE "\r\n");
      else if (status == ERR_NOT_IMPL)
	fprintf (ofile, "-ERR " NOT_IMPL "\r\n");
      else if (status == ERR_BAD_CMD)
	fprintf (ofile, "-ERR " BAD_COMMAND "\r\n");
      else if (status == ERR_BAD_LOGIN)
	fprintf (ofile, "-ERR " BAD_LOGIN "\r\n");
      else if (status == ERR_MBOX_LOCK)
	fprintf (ofile, "-ERR [IN-USE] " MBOX_LOCK "\r\n");
      else if (status == ERR_TOO_LONG)
	fprintf (ofile, "-ERR " TOO_LONG "\r\n");
      else if (status == ERR_FILE)
	fprintf (ofile, "-ERR " FILE_EXP "\r\n");
      else
	fprintf (ofile, "-ERR unknown error\r\n");

      free (buf);
      free (cmd);
      free (arg);
    }

  fflush (ofile);
  return (status != OK);
}

/* Runs GNU POP3 in standalone daemon mode. This opens and binds to a port
   (default 110) then executes a pop3d_mainloop() upon accepting a connection.
   It starts maxchildren child processes to listen to and accept socket
   connections.  */

void
pop3d_daemon (unsigned int maxchildren)
{
  SA server, client;
  pid_t pid;
  int listenfd, connfd;
  size_t size;

  listenfd = socket (AF_INET, SOCK_STREAM, 0);
  if (listenfd == -1)
    {
      syslog (LOG_ERR, "socket: %s", strerror(errno));
      exit (1);
    }
  size = 1; /* Use size here to avoid making a new variable.  */
  setsockopt (listenfd, SOL_SOCKET, SO_REUSEADDR, &size, sizeof(size));
  size = sizeof (server);
  memset (&server, 0, size);
  server.sin_family = AF_INET;
  server.sin_addr.s_addr = htonl (INADDR_ANY);
  server.sin_port = htons (port);

  if (bind (listenfd, (struct sockaddr *)&server, size) == -1)
    {
      syslog (LOG_ERR, "bind: %s", strerror (errno));
      exit (1);
    }

  if (listen (listenfd, 128) == -1)
    {
      syslog (LOG_ERR, "listen: %s", strerror (errno));
      exit (1);
    }

  for (;;)
    {
      if (children > maxchildren)
        {
	  syslog (LOG_ERR, "too many children");
          pause ();
          continue;
        }
      connfd = accept (listenfd, (struct sockaddr *)&client, &size);
      if (connfd == -1)
        {
          if (errno == EINTR)
	    continue;
          syslog (LOG_ERR, "accept: %s", strerror (errno));
          exit (1);
        }

      pid = fork ();
      if (pid == -1)
	syslog(LOG_ERR, "fork: %s", strerror (errno));
      else if (pid == 0) /* Child.  */
        {
	  int status;
          close (listenfd);
          status = pop3d_mainloop (connfd, connfd);
	  closelog ();
	  exit (status);
        }
      else
        {
          ++children;
        }
      close (connfd);
    }
}
